### PR TITLE
Add abstract_kernel and test

### PR DESCRIPTION
**Pull request by andrewstanfordjason**

### CODE REVIEW

File: lib_nn/api/AbstractKernel.hpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
+#pragma once
```
Comment (ghost) [inline lib_nn/api/AbstractKernel.hpp]: Could this be please changed to include guard?

File: lib_nn/api/AbstractKernel.hpp [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+     * The number of output channel groups that will be processed when this
+     * filter is executed.
+     */
+    int32_t output_channel_group_count;
+
+    // Used for setting the first channels slice, i.e. rather than writing to
+    // slice 0-31 by offsetting it we can address slice 32 - 63, etc.
+
+    /**
+     * The first output channel to be processed when this filter is executed.
```
Comment (ghost) [inline lib_nn/api/AbstractKernel.hpp:44]: Is this comment discussing something else?

**Review by ghost — APPROVED** [review]
After the comment changes are resolved, +1

File: lib_nn/api/AbstractKernel.hpp (hunk 2 above)
Comment (andrewstanfordjason) [reply lib_nn/api/AbstractKernel.hpp:44]: no, I'll merge them together

**Pull request merged**